docs/proposals/addon-auth-pattern.md

```diff
@@ -0,0 +1,122 @@
+
+The following pattern is in use today.
+
+When a cluster service, such as Dashboard is being deployed:
+1. Dashboard deployment and service are is created, in namespace `kube-system`.  Service is called `dashboard`.
+
+When a user wants to access a service:
+1. User runs `kubectl proxy -p 8001 &`.
+  -  This causes kubectl to serve HTTP on `localhost:8001`, and to reverse proxy that to the users current cluster, using TLS and the users
+     currently configured authentication method.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.
+  - The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/#/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. The Dashboard sees the request.  It does not know what users is making the request.
+1. The response is sent back from the Pod to the Apiserver.
+1. The response is sent back to the user.
+1. The user sees the Kubernetes Dashboard page.
+
+The problem with this is that the Dashboard does not know which user is accessing it, so it cannot support multiple users with different capabilities.
+
+### Proposed Pattern
+
+The proposed pattern has the following key improvements over the current pattern:
+- a header is injected which tells the Dashboard pod the user who is reaching it.
+- the header can be validated easily as being signed by the apiserver, which is trusted to authenticate users
+
+When a cluster service, such as Dashboard is being deployed:
+1. Dashboard deployment and service are is created, in namespace `kube-system`.  Service is called `dashboard`.
+  - *NEW* The service has annotation `alpha.service-proxy.kubernetes.io/proxy-authentication` set to `true`.
+    Only services that want the authentication header injected set this option.
+
+When a user wants to access a service:
+1. User runs `kubectl proxy -p 8001 &`.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.
+  - The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/#/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. *NEW* When the API server proxies a request to a service with annotation `alpha.service-proxy.kubernetes.io/proxy-authentication`, it does the following extra steps:
+    1. Injects a `X-Remote-User:` header with the verified username of the user. (Similar to [Openshift Request Header Authentication](https://docs.openshift.com/enterprise/3.0/admin_guide/configuring_authentication.html#RequestHeaderIdentityProvider) )
```

Comment:
Unless this header is signed or something, versions of components expecting that deal with this header would have to only be used with API servers supporting this.  Otherwise, if you used a such a component with an older version of the API server, a user could insert `X-Remote-User` of any value, and then API server would happily pass that along, including it's certificate (like it normally does).  Also, you'd have to make sure to sanitize the contents of X-Remote-User coming through the proxy, right?